### DISCUSSION TITLE
Ensuring HNSW Indexes Exist

### DISCUSSION THREAD

**Discussion by techieAgnostic** [cozodb/cozo#111] [Q&A] [answered]
Hello.

Is there an idiomatic way to ensure tables and indexes exist?
`:replace table { ... }` can be used with an empty relation, however I cannot find an equivilent for `::hnsw`.
My use case is with persistent storage, I would like to create indexes and tables if they do not already exist.

Thank you

**Comment by zh217** [comment, marked as answer]
We will work out a way. Note that `:replace` may not be what you want: it removes everything from the old relation if it exists!

Meanwhile you can `try ... catch ...` in the host language. If you use Python:

```python
try:
    for s in schema:
        db.run(s)
except QueryException as e:
    if e.code == 'eval::stored_relation_conflict':
        logger.warning('Database already initialized.')
    else:
        raise
```

For indices (every kind of indices), look for `e.code == 'tx::index_already_exists'` instead.

**Reply by andrewbaxter** [reply]
I was looking for an equivalent for `create if not exists` for schema handling and I think this is the answer.  Could this be added to the docs (for `:create` maybe?) and (Rust) is there an exported constant for `"eval::stored_relation_conflict"`?
